### ISSUE TITLE
Humble suggestion for code refactoring

### ISSUE THREAD

**Issue by kkonevets** [rusty1s/pytorch_geometric#1252]
https://github.com/rusty1s/pytorch_geometric/blob/5c22982a949b8c1597d1e8c13428adf4db0e45e2/torch_geometric/data/sampler.py#L186

Lines 186-198 could become this:

```python
            def set_edge_row(i, n_id, new_n_id):
                row_i = self.edge_index[i, e_id]
                if self.add_self_loops:
                    row_i = torch.cat([row_i, n_id], dim=0)
                self.tmp[new_n_id] = torch.arange(new_n_id.size(0))
                edges[i] = self.tmp[row_i]

            set_edge_row(self.i, n_id, n_id)
            set_edge_row(self.j, n_id, new_n_id)
```

it would be easier to read and understand

**Comment by rusty1s** [comment]
Thanks for your input. I completely revamped the `NeighborSampler` class. It is now much cleaner and way faster than before, even allowing for `num_workers` attributes to allow generation of subgraphs in parallel. Please have a look [here](https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.NeighborSampler).

**Comment by kkonevets** [comment]
Wow! I will definitely take a look in few days, thanks!

**Comment by kkonevets** [comment]
Is it possible to pass a `dtype` to `SparseTensor`? Because` from_edge_index()` takes my big `edge_index` array which does not fit in memory if I use `torch.long`. In my case I have 40Gb `edge_index` with `dtype=torch.int32` and I always do `***.to(torch.long)` when needed on batches. Now I can't do that. In 1.4.3 version I could modify `NeighborSampler` for my use case.

And I could't understand why do use a sparse tensor for `edge_index`? I thought it can't be sparse. But that does not matter a lot, just curious.

**Comment by kkonevets** [comment]
Edit:
Maybe you use sparse tensor for `edge_index` to access edges directly by node indexes, that could be convenient and fast, as apposed to finding a node in [2 X number_of_edges] edge_list. For example, you could use Adjacency matrix to save edges. Sorry for my guessing, just trying to understand.

**Comment by kkonevets** [comment]
And please explain the key difference between `pytorch_cluster/csrc/cpu/sampler_cpu.cpp` and `pytorch_sparse/csrc/cpu/sample_cpu.cpp`. In both you use unordered set to keep track of sampled items. What is the point to use later over the former?

**Comment by kkonevets** [comment]
Sorry for too many questions, I have one more.
It seems like `SAGEConv` has no concatenation now but summation instead, is it done on purpose, maybe has better test results?
 
![Screenshot from 2020-05-25 22-53-38](https://user-images.githubusercontent.com/16817116/82840297-af571380-9eda-11ea-859c-b6d6537502cf.png)

**Comment by rusty1s** [comment]
Sadly, there is no way to storage indices as `torch.int`, since PyTorch requires LongTensors for indexing. The `SparseTensor` is just a more convenient format as `edge_index`. It can additionally store `rowptr` or `rowcount`, but relies on sorted input. Therefore, we can easily find neighbors of specific nodes.

The key differences between the two implementations is that the latter supports `replace=False/True`, has faster sampling without replacement, and achieves same indexing for equal nodes. I have considered other sampling strategies (like shuffling), and they are all way slower.

Regarding `SAGEConv`, summation/concatenation is the same. Summation is better computation-wise though since it does not require to put node features into a new memory layout.

**Comment by kkonevets** [comment]
Thanks for your answers.
Yes, pytorch requires longs for indexing, but does not require to keep them as such, we can keep them as ints and convert to longs when needed. It is a crutial feature for a big graph. I guess I should do something with it on my own.

**Comment by kkonevets** [comment]
Or maybe I should just double my RAM, this could be easier :)
